### PR TITLE
Feature branch

### PR DESCRIPTION
**Pull request by cristianCeamatu**
This is a private events app following the project requirements from TheOdinProject website. The project enforced our skills in using Model associations and create from scratch Models, Controllers, Views, and Migrations.

**App functionality**
- Users can be created, they can sign in and sign out
- Users can create and attend events (they cannot attend events they are already attending)
- Events are categorized in past and upcoming events based on their date

### CODE REVIEW

File: app/models/event.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class Event < ApplicationRecord
+  has_many :event_attendees, foreign_key: :attended_event_id
+  has_many :attendees, through: :event_attendees, source: :event_attendee
+  belongs_to :creator, class_name: 'User'
+
+  scope :upcoming_events, -> { where('date > ?', Date.today) }
```
Comment (addod19) [inline app/models/event.rb]: Kindly replace `'date > ?'` with `'date >= ?'`

**Review by addod19 — CHANGES_REQUESTED** [review]
## Status: Require changes :recycle: :traffic_light: 

Hi, @cristianCeamatu and @uimarshall, great job with this project :+1: README is quite detailed, PR is professional, has a title and a short description:smile: KIndly address the comments below so we get our project approved :+1: 
- Let us add the migration command before we run `rails s`. Otherwise, we will run into the `Pending migration error`
- We are currently missing the `foreign keys` in our migration file. Let us check our migration files so we can fix them. Kindly check out this [link](
https://api.rubyonrails.org/v5.1.7/classes/ActiveRecord/ConnectionAdapters/TableDefinition.html#method-i-references) or [this](
https://api.rubyonrails.org/v5.1.7/classes/ActiveRecord/ConnectionAdapters/TableDefinition.html#method-i-belongs_to) in adding foreign keys
- `Events Controller`, our `Index` action is missing the following;
  - [x] `@past_events = Event.previous_events`
  - [x] `@upcoming_events = Event.upcoming_events`
- `Users Controller`, our `Show` action is missing the following:
  - [x] List events created by a user `@created_events = current_user.events`
  - [x] List event previously attended by a user ` @past_events = current_user.attended_events.previous `
  - [x] List upcoming events previously attended by a user using event model scope `
@upcoming_events = current_user.attended_events.upcoming`
- `Test`, currently, the `rspec` command as outlined in the README file is not working. Kindly run `rails db:migrate RAILS_ENV=test` before running the `rspec`
We currently have some test failing, kindly check the screenshot below to understand the issue
ftest
- Events created on the same day is not being recorded, kindly fix that in the `event scope model`

Kindly address the issue above, once these are fixed, kindly submit another code review :slightly_smiling_face: 
_Happy coding_ :man_technologist: 
[Daniel Larbi Addo](linkedin.com/in/daniel-larbi-addo/)

**Comment by cristianCeamatu** [comment]
@addod19  Thank you for a very constructive code review.
Everything updated as per requests, have a great day!

File: app/views/users/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                <li class="list-group-item">- <strong>Date:</strong> <%= event.date %></li>
+                <li class="list-group-item">- <strong>Location:</strong> <%= event.location %></li>
+                <li class="list-group-item">- <strong>Description:</strong> <%= event.description %></li>
+              </ul>
+            </div>
+          </div>
+        <% end %>
+      <% else %>
+        <p>No upcoming events by <%= @user.name %></p>
+      <% end %>
```
Comment (maelfosso) [inline app/views/users/show.html.erb]: Logic not allowed in views. Kindly implement them in views helpers.

File: app/views/users/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                  <li class="list-group-item">- <strong>Date:</strong> <%= event.date %></li>
+                  <li class="list-group-item">- <strong>Location:</strong> <%= event.location %></li>
+                  <li class="list-group-item">- <strong>Description:</strong> <%= event.description %></li>
+                </ul>
+              </div>
+            </div>
+          <% end %>
+        <% else %>
+          <p>No past events by <%= @user.name %></p>
+        <% end %>
```
Comment (maelfosso) [inline app/views/users/show.html.erb]: Logic not allowed in views. Kindly implement them in views helpers.

File: app/views/users/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        <h3>Past events</h3>
+        <% if @past_events.any? %>
+          <% @past_events.each do |event| %>
+            <p><%= event.date %> - <%= event.location %> - <%= event.description %></p>
+          <% end %>
+        <% else %>
+          <p>No past events</p>
+        <% end %>
+      </div>
+    </section>
```
Comment (maelfosso) [inline app/views/users/show.html.erb]: Logic not allowed in views. Kindly implement them in views helpers.

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -1,15 +1,42 @@
...
+    <nav class="navbar navbar-expand-lg navbar-light bg-primary mb-0">
+      <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
+        <span class="navbar-toggler-icon"></span>
+      </button>
+      <div class="collapse navbar-collapse" id="navbarSupportedContent">
+        <ul class="navbar-nav mr-auto">
+          <li><%= link_to 'Private Events', root_path, class: 'text-white font-weight-bold' %></li>
+        </ul>
+        <form class="form-inline my-2 my-lg-0">
+          <% if !session[:user_id].nil? %>
```
Comment (maelfosso) [inline app/views/layouts/application.html.erb]: Logic not allowed in views. Kindly implement them in views helpers.

Comment (maelfosso) [reply app/views/layouts/application.html.erb]: It's preferable to create a helper called `current_user` instead of using `session[:user_id]`

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Thank you for applying the suggested changes.
However, some requirements are missings

1- You've forgotten to add indexes in `Attendance` table

2- Kindly remove ogic from views. Read the comment leff below

3- For the unit tests, I suggest you to use the **shoulda-matcher** gem which is really helpful

Kindly make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -1,15 +1,42 @@
...
+          <li><%= link_to 'Private Events', root_path, class: 'text-white font-weight-bold' %></li>
+        </ul>
+        <form class="form-inline my-2 my-lg-0">
+          <% if current_user %>
+            <%= link_to "Signed In as " + @user.name, user_path(@user),class: 'text-white ml-3'  %>
+            <%= link_to 'All events', events_path, class: 'btn btn-light my-2 my-sm-0 text-info ml-2' %>
+            <%= link_to 'Create event', new_event_path, class: 'btn btn-light my-2 my-sm-0 text-info ml-3' %>
+            <%= link_to "Sign out", sign_out_path, method: :delete, class: 'btn btn-light my-2 my-sm-0 text-info ml-2' %>
+          <% else %>
+            <%= link_to 'All events', events_path, class: 'btn btn-light my-2 my-sm-0 text-info ml-2' %>
```
Comment (maelfosso) [inline app/views/layouts/application.html.erb:33]: There are still logic in your code. please, move them into views helpers

File: app/views/events/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+          <%= link_to event_path(event) do %>
+          <ul class="list-group list-group-flush">
+            <li class="list-group-item">- <strong>Date:</strong> <%= event.date %></li>
+            <li class="list-group-item">- <strong>Location:</strong> <%= event.location %></li>
+            <li class="list-group-item">- <strong>Description:</strong> <%= event.description %></li>
+          </ul>
+          <% end %>
+        </div>
+
+        <% if current_user && !@user.attended_events.map { |el| el.id }.include?(event.id) %>
```
Comment (maelfosso) [inline app/views/events/index.html.erb]: That is first logic in views and it must be moved into view helpers. Secondly, that is a request into views which is not allowed.
Move that into  `Event` model and call it after inside your view helper

File: spec/models/event_spec.rb [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    #   @event.place = 'xxx'
+    #   @event.save
+    #   expect(@event).to_not be_valid
+    # end
+
+    # it 'should be valid if event:date is nil' do
+    #   @event.date = nil
+    #   @event.save
+    #   expect(@event).to_not be_valid
+    # end
```
Comment (maelfosso) [inline spec/models/event_spec.rb]: Remove the comments

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello, 

Thank you for applying the suggested changes.
However, you missed some.
Kindly read the comment left below, make the suggested changes and submit a new code review request

Thanks

Mael FOSSO

File: app/views/layouts/application.html.erb (hunk 6 above)
Comment (cristianCeamatu) [reply app/views/layouts/application.html.erb:33]: All the tutorials and reading materials we followed up used if/else statements in their views, also the #any? and #each methods. This is the first time we see that you cannot use IF/ELSE statements in your view. Below is a screenshot of an official RAILCASTA video where they are doing exactly what we are doing in our project.

**Review by wathigo — APPROVED** [review]
### Approved :heavy_check_mark: 
Great job on this project.

**You may move on.**

**Good Luck :four_leaf_clover: 
Simon Wathigo.**

**Pull request merged**